algolia/algoliasearch-sfcc-b2c
# feat: Base Product level indexing

## Base Product level indexing

Introduce an option to index products at the base product level:
It creates one record per base product, with all variants embedded in it.

### Changes

Naming:
- Use camelCase for all new attributes: `color_variations` has been renamed to `colorVariations`

Job:
- new `RecordModel` configuration parameter
- new `jobHelper.generateMasterRecords()` function, which builds the localized Algolia records from a master product
- `algoliaProductIndex` job now uses the `RecordModel` parameter, and call the `generateMasterRecords()` helper for base product indexing
- throw an error if the indexingMethod is `partialRecordUpdate` and the attributes belong to variants: we can't partially update objects in an array

Frontend:
- Rely on the `RecordModel` parameter to know which facet to apply (for size and color)
- For each variant:
  - Add the tracking parameters (queryID, indexName, ...) to the variant URL
  - Select the correct variant to display: when one of the variants is matching the search criterias, the whole base product record is returned. The frontend needs to chose which variant to display. For that, we use the `disjunctiveFacetsRefinements` present in InstantSearch state
  - Select which image to display: to avoid records getting too big, images are stored at the base product level in the `colorVariations` attribute. Select the image corresponding to the selected variant
  - Use the price and URL of the variant

Tests:
- Added test for the new `generateMasterRecords` helper
- Added tests for the jobs with `master-level` record model

### How to test

- Select 'Base Product' as the record model in the Business Manager
- Run the AlgoliaProductIndex_v2 in `fullCatalogReindex` (to avoid having old variant-based records in the index). Or use a different index prefix
- The UI should have the same capabilities than for variant-level records: faceting, swatches, ...


---
SFCC-239
